engine/crates/engine-v2/src/request.rs

```diff
@@ -0,0 +1,152 @@
+use engine::{ServerError, ServerResult};
+pub use engine_parser::types::OperationType;
+use engine_parser::Positioned;
+use schema::{FieldId, Schema};
+
+use crate::response::{Argument, OperationSelection, OperationSelectionSet, ResponseFields, ResponseFieldsBuilder};
+
+#[derive(Clone, Copy, PartialEq, Eq, Hash)]
+pub struct VariableId(usize);
+
+pub struct OperationDefinition {
+    pub ty: OperationType,
+    pub selection_set: OperationSelectionSet,
+    pub response_edges_builder: ResponseFieldsBuilder,
+}
+
+pub struct OperationBinder<'a> {
+    response_edges_builder: ResponseFieldsBuilder,
+    schema: &'a Schema,
+}
+
+impl<'a> OperationBinder<'a> {
+    pub fn new(schema: &'a Schema) -> Self {
+        Self {
+            response_edges_builder: ResponseFields::builder(),
+            schema,
+        }
+    }
+
+    pub fn bind(mut self, operation: engine_parser::types::OperationDefinition) -> ServerResult<OperationDefinition> {
+        let root_object_id = match operation.ty {
+            OperationType::Query => self.schema.root_operation_types.query,
+            OperationType::Mutation => self
+                .schema
+                .root_operation_types
+                .mutation
+                .expect("Mutation operation type not supported by schema."),
+            OperationType::Subscription => self
+                .schema
+                .root_operation_types
+                .subscription
+                .expect("Subscription operation type not supported by schema."),
```

Comment:
This function returns a Result, so these should probably be `ok_or_else(...)?`

Reply:
oh yeah definitely. I'll need to make a first pass over the code to add some error management which is pretty much nonexistent currently. ^^